Trigger Home Assistant shutdown automations right before the stop event instead of during it

tetele opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The event key in the (undocumented) trigger data for the `homeassistant` trigger platform is now the string `shutdown` for a `homeassistant` trigger configured to fire on shutdown instead of an `Event` object.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR is an alternative to #83781 and a result of this architectural discussion https://github.com/home-assistant/architecture/discussions/827. When having automations triggered by the Home Assistant shutdown event, some integrations (notably, the Recorder) stop before any entity-changing service calls get processed. As a result, these automations 
don't get the entity states persisted.

https://github.com/home-assistant/core/pull/91260 addresses part of this issue by moving the Recorder unloading until the `FINAL_WRITE` stage, but still some other integrations required by the automations might be unloaded before the automations themselves run.

The change ([proposed by balloob](https://github.com/home-assistant/core/issues/41396#issuecomment-704969818)) splits the `EVENT_HOMEASSISTANT_STOP`, activating the `shutdown` trigger of the `homeassistant` integration a step before the integrations stopping.

To do so, part of the time allocated for the integrations stopping (20s out of 100s) was instead allocated to the automations to run. The remaining 80s should be enough to allow other integrations to shut down. Developer docs updated in this PR https://github.com/home-assistant/developers.home-assistant/pull/1744


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #89109, #79919, #76765, #73081, #59160, #41396, [these](https://community.home-assistant.io/t/automation-on-homeassistant-shutdown-event-no-longer-working/204282) [issues](https://community.home-assistant.io/t/configuring-an-automation-trigger-for-hass-shutdown/98708/6?u=tetele) on the [community forum](https://community.home-assistant.io/t/ha-shutdown-automations-dont-work/232729) and others
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26987

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tetele committed:
Shutdown fix

tetele committed:
Fix timeout to not overpass supervisor timeout

test_cumulative_shutdown_timeout_less_than_supervisor

bdraco commented on homeassistant/core.py:
```diff
@@ -760,8 +762,8 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
         # stage 2
-        self.state = CoreState.final_write
-        self.bus.async_fire(EVENT_HOMEASSISTANT_FINAL_WRITE)
+        self.state = CoreState.stopping
```
Isn't this already set above?

bdraco commented:
The PR title is a bit confusing since this change doesn't seem to be affecting the recorder directly

tetele committed:
Avoid core state overwrite

bdraco commented on homeassistant/core.py:
```diff
@@ -99,9 +100,10 @@
-STAGE_2_SHUTDOWN_TIMEOUT = 60
-STAGE_3_SHUTDOWN_TIMEOUT = 30
+STAGE_1_SHUTDOWN_TIMEOUT = 60
+STAGE_2_SHUTDOWN_TIMEOUT = 40
```
Lowering the timeout for STOP is likely a breaking change as it mean integration only have 40 seconds to react to a stop event

bdraco replied:
I'd rename these constants to match the corresponding name of the event since we now have a lot of stages

tetele replied:
You're right. Plus, unloading is more important than automations.

I think I should also change the log messages in case of timeouts, not just the constants. Thoughts?

bdraco replied:
Would make sense to adjust the log messages as well

Also it's very important that the total amount of time in these constants doesn't increase as core could get killed by supervisor if it takes too long since it will think it's crashed

bdraco replied:
Also any adjustment downward has to be a breaking change since it affects the maximum time an integration has to shutdown. That's probably ok though

bdraco commented:
> I thought the passive voice ("being shut down") was enough, but i can rephrase it. I just wanted to avoid explaining the code in the title and focus on the outcome.
> 
> PS: sorry for messing up the PR merge earlier and dragging 4000 reviewers along for the ride.

I was thinking something more like "Trigger Home Assistant shutdown automations before the stop event instead of during it"

tetele committed:
More explicit stage names

Edit the stage names and associated log messages to better understand what's going on

bdraco commented on homeassistant/core.py:
```diff
@@ -99,9 +100,10 @@
-STAGE_1_SHUTDOWN_TIMEOUT = 100
-STAGE_2_SHUTDOWN_TIMEOUT = 60
-STAGE_3_SHUTDOWN_TIMEOUT = 30
+AUTOMATION_STAGE_SHUTDOWN_TIMEOUT = 20
```
Lots call this stop pending to match the event name since something else might want to use that

tetele committed:
Rename "stop pending" stage

tetele committed:
Fix running tasks log messages

bdraco commented on homeassistant/core.py:
```diff
@@ -813,14 +829,14 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
         except asyncio.TimeoutError:
             _LOGGER.warning(
-                "Timed out waiting for shutdown stage 3 to complete, the shutdown will"
+                "Timed out waiting for Home Assistant core integrations to stop, the shutdown will"
```
Please name the message after the event so any bug reports can make it easier to figure out which event we are waiting for 
```suggestion
                "Timed out waiting for close event to be procssed, the shutdown will"
```

Also there is no need to mention Home Assistant core since this will already be logged as `homeassistant.core`

tetele committed:
Rephrase close event log message

bdraco commented on homeassistant/core.py:
```diff
@@ -748,31 +750,43 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
         except asyncio.TimeoutError:
             _LOGGER.warning(
-                "Timed out waiting for shutdown stage 1 to complete, the shutdown will"
+                "Timed out waiting for shutdown automations to run, the shutdown will"
```
Let's call this by the name of the event since other integrations can use it

tetele committed:
Rephrase stop pending log message

tetele commented:
@bdraco I've actually added some documentation stating the timeout for automations to run https://github.com/home-assistant/home-assistant.io/pull/26954

Why the `docs-missing` label? Should I add anything else to the docs?

tetele commented:
I will try to add to the dev docs as well.

Why do you say there's a bug in the recorder? I think it's just doing its job, unloading before HA gets shut down. How else would you see it working?

bdraco commented:
> Why do you say there's a bug in the recorder? I think it's just doing its job, unloading before HA gets shut down. How else would you see it working?

https://github.com/home-assistant/core/pull/91260

epenet commented:
I am following this discussion as I recently added `self._cancel_cancellable_timers()` and I keep thinking that maybe I should have moved it down until after the STOP event.
```python
        self.state = CoreState.stopping
        self.bus.async_fire(EVENT_HOMEASSISTANT_STOP)

        self._cancel_cancellable_timers()

        try:
            ...
```

From what I can see, what you are trying to solve on the automation trigger potentially applies to all integrations

bdraco commented:
We could also do both. Its getting late here. I'll sleep on this one and see if I have any more ideas tomorrow.

tetele commented:
@bdraco I see that your PR that moves the recorder shutdown on `FINAL_WRITE` got approved. Although that sort of patches the initial issue (i.e. automations that change states not storing the state changes), they might still fail because an integration that they use gets unloaded before they actually run.

I will change the description of this PR and I really think that we should have an extra step for running automations.

bdraco commented:
> @bdraco I see that your PR that moves the recorder shutdown on `FINAL_WRITE` got approved. 

I still need to write some tests for that and spend a few more hours validating the behavior is a good change.  I think it is, but I want to spend a few more hours testing how it interacts and write tests to validate assumes before moving it forward.

> Although that sort of patches the initial issue (i.e. automations that change states not storing the state changes), they might still fail because an integration that they use gets unloaded before they actually run.

That makes sense.  From a functional standpoint this PR looks good to me. As its a significant change to core that affects a lot of people and integrations, we should wait for additional feedback before moving it forward.

> 
> I will change the description of this PR and I really think that we should have an extra step for running automations.

👍

tetele commented:
What exactly is the workflow for an issue with the label `second-opinion-wanted` but with no other reviews requested? I can't ask for reviews from anyone. Is there anything I should do?

bdraco commented:
> What exactly is the workflow for an issue with the label `second-opinion-wanted` but with no other reviews requested? I can't ask for reviews from anyone. Is there anything I should do?

Someone will pick it up when they are doing reviews. Short version is.. waiting

edenhaus reviewed (dismissed):
Please rebase your PR and verify that the CI is passing.

edenhaus commented:
@tetele Can you please add some tests for your changes? We require at least 97% of the patch must be covered by tests.

tetele commented:
@edenhaus I've already added tests. From what I see, the lines lacking coverage are the logging instructions, which lacked the coverage even before my patch. Should those really be tested?

tetele commented:
@edenhaus should I keep updating this PR? Should I close it? What should I do regarding the tests for log messages, as I asked in the previous comment?

edenhaus commented:
If you have time, also add tests for the logging :)

Please don't close the PR. Only be patient as more than 500 PRs are open and we do our best to review all.
Before merging, we still need a second opinion from the core team.

tetele commented:
Just leaving this here for future reference: code coverage should not be an issue, as this message states https://discord.com/channels/330944238910963714/554842238073700352/1140665680836956282

edenhaus commented:
@tetele Only if it is not needed, does not mean you should not implement it if you don't have the time :)

tetele commented:
Yeah, duly noted, but no, thanks!

Last time I tried writing tests that detected events being interrupted, it took me [3 days and 1 moderator](https://discord.com/channels/330944238910963714/1149683444352286854) only to find out I've written them [in vain](https://github.com/home-assistant/core/pull/99870#pullrequestreview-1618897305). Not to mention this particular PR required a fair number of updates from me and it's still pending after 5 months.

I'd rather wait until I get a merged feature PR, 'cause otherwise I feel I'm just wasting my time. [Any](https://github.com/home-assistant/core/pulls/tetele) merged PR.

frenck reviewed (dismissed):
I think this PR makes tons of sense and looks good at first glance.

But... this does add stages to the Core engine, this really needs tests.

../Frenck

tetele committed:
Rename event according to feedback

tetele committed:
Some more smaller renames

joostlek commented:
An idea I might be having would be to patch the TIMEOUT variables to something low, and then try to have a job in the executor that sleeps longer than the set timeout. This should trigger the timeout.

We don't like touching internals, but not sure if this is an exception.

(Maybe @MartinHjelmare has an idea on how to improve tests here?)

erwindouna commented on tests/test_core.py:
```diff
@@ -386,56 +387,66 @@ def run(self) -> None:
     assert len(test_stop) == 1
     assert len(test_close) == 1
     assert len(test_final_write) == 1
-    assert len(test_all) == 2
```
Try this:

`with patch.object(hass.timeout, "async_timeout", side_effect=asyncio.TimeoutError):
        await hass.async_stop()

    assert hass.state == CoreState.stopped`

erwindouna commented on tests/test_core.py:
```diff
@@ -386,56 +387,66 @@ def run(self) -> None:
     assert len(test_final_write) == 1
-    assert len(test_all) == 2
+    assert len(test_all) == 3
 
```
```with patch.object(hass.timeout, "async_timeout", side_effect=asyncio.TimeoutError):
        await hass.async_stop()

    assert hass.state == CoreState.stopped
```

erwindouna commented on tests/test_core.py:
```diff
@@ -386,56 +387,66 @@ def run(self) -> None:
     assert len(test_final_write) == 1
-    assert len(test_all) == 2
+    assert len(test_all) == 3
 
```
```
with patch.object(hass.timeout, "async_timeout", side_effect=asyncio.TimeoutError):
        await hass.async_stop()

    assert hass.state == CoreState.stopped
```

erwindouna replied:
Try adding this below. On my test setup this gets rids of the missing TimeoutErrors from your code coverage.

tetele committed:
Attempt to fix code coverage issues

erwindouna commented:
That seemed to work. An increase of 20%, for your patch. One line remaining. Let me try to figure that out a bit later.

tetele commented:
@erwindouna you're the man! Many, many thanks!

erwindouna commented:
@bdraco There's one unit test left. It for the Exception in core.py on line 892.
We can't mock `hass._tasks`, since later it will be overwritten by `running_tasks`. Thus a side effect on task.cancel doesn't work.
Any ideas how we can help tetele on triggering this exception properly?
We're almost there and I also seem to be overlooking this.

tetele committed:
Extract timeouts as a separate test

tetele commented:
OK, I've still got an issue with [line 892](https://github.com/home-assistant/core/blob/4a5cab27f6a579e4bbbb00c8a976c1327cb6a8fc/homeassistant/core.py#L892), where my only change was to make the error message more specific.

I can't figure out how to reach that line. To me it looks like a safeguard and it shouldn't ever be reached.

I've tried the following tests:
| **Approach** | **Cause of failure** |
| --- | --- |
| hooking into `EVENT_HOMEASSISTANT_CLOSE` to add a task, then mock that task's `cancel` method to fail | the running tasks [are moved over](https://github.com/home-assistant/core/blob/4a5cab27f6a579e4bbbb00c8a976c1327cb6a8fc/homeassistant/core.py#L809) to a local variable called `running_tasks` at the beginning of the shutdown process |
| patch `asyncio.Future.cancel` to selectively raise a generic exception shortly after the task is cancelled | I can't patch an immutable type `asyncio.Future`. Also tried with `asyncio.Task` |
| questioning whether we need [that `except Exception` branch](https://github.com/home-assistant/core/blob/4a5cab27f6a579e4bbbb00c8a976c1327cb6a8fc/homeassistant/core.py#L891) at all | you can catch the `CancelledError` inside the wrapped coro and the task gets uncancelled, so we need to safeguard for that. Still thinking about this |
| adding a task before they get copied to the local variable `running_tasks`. The task waits for an `asyncio.Event` that never gets `set()`, while its `cancel` method is patched to wait for 0.05s (half of [the timeout](https://github.com/home-assistant/core/blob/4a5cab27f6a579e4bbbb00c8a976c1327cb6a8fc/homeassistant/core.py#L882C30-L882C30)), then raise a generic Exception | `Detected blocking call to sleep inside the event loop` |

I can't thank @erwindouna enough for all the provided help ❤️ 

Can anyone from the @home-assistant/core team please advise? My 2 main questions are:
- is that generic `Exception` branch really needed?
- what would be the best approach to test that an exception raised by a task in the process of being cancelled does not prevent the shutdown

tetele committed:
Hopefully finished code coverage

erwindouna commented:
Seems fixed now. :)

emontnemery commented on homeassistant/core.py:
```diff
@@ -820,31 +822,43 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
 
         # stage 1
         self.state = CoreState.stopping
```
Why don't we introduce a new core state too? Also, shouldn't we set the state to `CoreState.stopping` just before firing `EVENT_HOMEASSISTANT_STOP` and not 20s before that?

tetele replied:
You're right, but the renaming of the concerned event was [a recent development](https://github.com/home-assistant/core/pull/91165/commits/27b6b6cbfec44881a98af9e426b58ca288450ee6#r1379937259), which made me forget that there was a `CoreState` named that way which could cause a confusion.

emontnemery reviewed (dismissed):
I agree this PR makes a lot of sense, and the implementation is essentially fine.

Some comments and questions though.

tetele commented on homeassistant/core.py:
```diff
@@ -108,9 +109,10 @@
-STAGE_1_SHUTDOWN_TIMEOUT = 100
-STAGE_2_SHUTDOWN_TIMEOUT = 60
-STAGE_3_SHUTDOWN_TIMEOUT = 30
+STOPPING_STAGE_SHUTDOWN_TIMEOUT = 20
+STOP_STAGE_SHUTDOWN_TIMEOUT = 80
+FINAL_WRITE_STAGE_SHUTDOWN_TIMEOUT = 60
+CLOSE_STAGE_SHUTDOWN_TIMEOUT = 30
```
I renamed the stages as per [bdraco's request](https://github.com/home-assistant/core/pull/91165/commits/eaf6d024bcb05cb4fecb90af487cda9256b74b57#r1163268933). This renaming, accompanied by the editing of the log messages, has caused my [code coverage to plummet](https://github.com/home-assistant/core/pull/91165#issuecomment-1798656458), leaving me unable to write tests due to lack of Python skills until recently, with aid.

Now you want me to change the stage names back. Is that correct?

tetele replied:
@emontnemery do you want me to rewrite the whole thing again to rename the events, timeouts etc.?

emontnemery replied:
I didn't know @bdraco requested that, I'll check with him.

Also, how can renaming some constants cause code coverage to plummet? Or you mean it plummets for this PR because you change already untested code?

tetele replied:
> Or you mean it plummets for this PR because you change already untested code?

Exactly

tetele commented on homeassistant/core.py:
```diff
@@ -108,9 +109,10 @@
-STAGE_2_SHUTDOWN_TIMEOUT = 60
-STAGE_3_SHUTDOWN_TIMEOUT = 30
+STOPPING_STAGE_SHUTDOWN_TIMEOUT = 20
+STOP_STAGE_SHUTDOWN_TIMEOUT = 80
```
Also [a suggestion](https://github.com/home-assistant/core/pull/91165/commits/eaf6d024bcb05cb4fecb90af487cda9256b74b57#r1163268235) from bdraco, who led me to believe that altering the supervisor is [out of the question](https://github.com/home-assistant/core/pull/91165/commits/eaf6d024bcb05cb4fecb90af487cda9256b74b57#r1163279316).

Even if it's fine to alter the supervisor, I will surely not contribute to that due to (again) lack of skills for writing tests.

emontnemery commented:
@tetele I discussed this PR with @bdraco. If you don't mind I'll rebase it to fix the merge conflict and also do some very minor adjustments to it? Or would you rather finish the PR yourself?

emontnemery committed:
Restore stop stage timeout

emontnemery committed:
Avoid prematurely changing states

emontnemery commented:
Needs supervisor PR https://github.com/home-assistant/supervisor/pull/4736

emontnemery committed:
Update Dockerfile

emontnemery commented:
I discussed with @bdraco, we don't want to merge this until after HA Core 2023.12 beta is cut, meaning it will be included in HA Core 2024.1

emontnemery commented:
I wonder if the implementation of this PR is fundamentally flawed.

The rationale for introducing yet another shutdown stage is to give automations triggered by shutdown 20s to run unimpeded.
However, we cancel background tasks before that, and the way to decide if the stage succeeded is to wait until all tasks are finished.

Should we instead:
- Shutdown background tasks only after stage 1 has finished
- Determine if shutdown succeeded by waiting for automations started by the `homeassistant` trigger to finish.

emontnemery commented:
> The actual beta testing will take just as long next month

Yes, but the implementation will be available on the dev branch in daily builds for a full month, which is the point of waiting

tetele commented:
At the moment, NO automations are guaranteed to work if triggered by the shutdown event. Some may, but if another platform/component that they require has been unloaded, then they won't.

So it's not like anything new could break :)

tetele commented:
Btw, sometimes a HA restart takes 2 minutes for me because i have automations triggered by shutdown. Sure, the obvious choice is to disable them, but that means that the shutdown trigger is utterly useless at this point.

emontnemery committed:
Only wait for automations started by shutdown trigger

emontnemery commented:
Commit ef4773466e37a0bc6c7f1e4efb3a27fd84d1315c is a suggestion to make stage 1 only about waiting for automations and nothing else

emontnemery commented:
> At the moment, NO automations are guaranteed to work if triggered by the shutdown event. Some may, but if another platform/component that they require has been unloaded, then they won't.

> So it's not like anything new could break :)

The point is that something we didn't intend to change may very well break.

emontnemery commented on homeassistant/components/homeassistant/triggers/homeassistant.py:
```diff
@@ -46,8 +46,11 @@ def hass_shutdown(event):
             )
+            if not fut_or_none:
+                return
+            hass.data.setdefault("homeassistant_triggers", []).append(fut_or_none)
```
The magic string is not nice, just used here to quickly illustrate the concept of being able to wait only for automations triggered by shutdown.

emontnemery committed:
Update _generate_dockerfile

emontnemery committed:
Rewrite shutdown trigger according to discussion

emontnemery commented on homeassistant/components/homeassistant/triggers/homeassistant.py:
```diff
@@ -30,24 +30,17 @@ async def async_attach_trigger(
+                "trigger": {
+                    **trigger_data,
+                    "platform": "homeassistant",
+                    "event": event,
```
I'm not sure this is valid, does it have to be an event object?

emontnemery replied:
Trigger data is documented here: https://www.home-assistant.io/docs/automation/templating#available-trigger-data, but the trigger data for the `homeassistant` trigger is undocumented. I still think this should be considered a breaking change though, in case someone relies on it.

bdraco replied:
Marking this resolved as its labeled a breaking change now

emontnemery commented:
I discussed the implementation a bit with @bdraco. Instead of having core know about homeassistant triggers, the latest commit adds function `hass.async_add_shutdown_job` which registers a job to be run when Home Asssistant is shutting down.

bdraco commented on homeassistant/core.py:
```diff
@@ -768,6 +780,42 @@ async def _await_and_log_pending(
+        job_with_args = HassJobWithArgs(hassjob, args)
+        self._shutdown_jobs.append(job_with_args)
+
+        @callback
+        def remove_job() -> None:
+            self._shutdown_jobs.remove(job_with_args)
+
+        return remove_job
```
Should these be removable?  I think we are past the point of no return on these that removal doesn't seem like it would be useful

emontnemery replied:
> we are past the point of no return on these that removal doesn't seem like it would be useful

I don't follow. If an automation is set up to trigger on shutdown, and the automation is modified or disabled, the trigger should be removed.

bdraco commented on homeassistant/core.py:
```diff
@@ -818,33 +887,32 @@ async def async_stop(self, exit_code: int = 0, *, force: bool = False) -> None:
                 " continue"
             )
-            self._async_log_running_tasks(1)
+            self._async_log_running_tasks(2)
```
Since we renamed the stages from `1`, `2`, etc, `1`,`2`, should be named as well so they match the stage name

emontnemery replied:
Makes sense, fixed

emontnemery committed:
Remove EVENT_HOMEASSISTANT_STOPPING

emontnemery committed:
Improve comments

emontnemery committed:
Improve debug logs

emontnemery committed:
Fix tests

emontnemery committed:
Add tests of async_add_shutdown_job

bdraco commented:
Single restart went ok.

I restart about 20-50 times per day so it will get tested in the normal course for other things

frenck reviewed (approved):
This PR looks absolutely great to me.

I think we can merge this and continue testing in nightly.

../Frenck

tetele commented:
I've never thought I'd live the day 😅 Thank you, gentlemen!

bdraco commented:
I'll update another production instance shortly, approve, and merge if everything goes ok

bdraco commented:
Shutdown stalled a bit but may be unrelated.  digging in

bdraco commented:
I can't get it to happen again after 50 more restarts 😢

bdraco commented:
Got it to happen again. The problem is unrelated.

I think this is good to go.

bdraco commented:
Restarted a few more times for testing bluetooth and everything is going well so far.

Thanks @tetele @emontnemery @frenck for all your contributes and reviews on this one. I know its been a bit of a journey on this one, as changing core flow almost always is.
